### PR TITLE
Support for expressions

### PR DESCRIPTION
**Pull request by dorantor**
How it's used in our project(simplified):
```php
    $uuid = '0f372656-6a5b-4727-a4c4-f6357775d926';
    $uuidExpr = $chc->expr("UUIDStringToNum('{$uuid}')");
    $stat = $chc->insert(
        'sometable',
        ['2018-10-18', $uuidExpr],
        ['created_date', 'event_id',]
    );
```

Where simplified `sometable` is:
```sql
CREATE TABLE sometable (
    created_date Date,
    event_id FixedString(16)
) ENGINE = MergeTree(
    created_date,
    event_id,
    (created_date, event_id),
    8192
);
```

### CODE REVIEW

File: src/Query/Expression.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * Query expression
+ *
+ * @package ClickHouseDB
+ */
+class Expression
+{
+    /**
+     * @var string
+     */
+    protected $expression = '';
```
Comment (simPod) [inline src/Query/Expression.php]: IMO there should not be a default value and not be protected. Let there only be a constructor as the way to create a valid object.
```suggestion
    private $expression;
```

File: src/Query/Expression.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * Expression constructor.
```
Comment (simPod) [inline src/Query/Expression.php]: This comment can be removed

File: src/Query/Expression.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace ClickHouseDB\Query;
+
+/**
+ * Query expression
```
Comment (simPod) [inline src/Query/Expression.php]: This would deserve more explanation

File: src/Query/Expression.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * @package ClickHouseDB
```
Comment (simPod) [inline src/Query/Expression.php]: This comment can be removed

File: src/Client.php [hunk 5]
```diff
@@ -888,4 +889,16 @@ public function dropOldPartitions(string $table_name, int $days_ago, int $count_
...
     }
+
+    /**
+     * Build sql expression to be used "as is" in a query
+     *
+     * @param string $expression
+     *
+     * @return \ClickHouseDB\Query\Expression
+     */
+    public function expr($expression)
```
Comment (simPod) [inline src/Client.php]: Do we need this method? I think that instantiating the Expression class is enough. WDYT?

File: src/Query/Expression.php (hunk 2 above)
Comment (dorantor) [reply src/Query/Expression.php]: phpdoc should be on every method IMHO, but it's your code, so it's up to you to decide.

Comment (dorantor) [reply src/Query/Expression.php]: Actually, I can see phpdoc at __constructor for `ClickHouseDB\Query\Query`. Are you sure it should be removed? :)

File: src/Client.php (hunk 5 above)
Comment (dorantor) [reply src/Client.php]: For convenience, mostly.

File: src/Query/Expression.php (hunk 2 above)
Comment (simPod) [reply src/Query/Expression.php]: It's @smi2's code actually ;) I just try to contribute especially in direction of cleanup and refactorings.

But there's no reason to say it's constructor with PHPDocs when code says that for you

```php
class Expression {
    public function __constructor
```
already says the same as comment `* Expression constructor.`

You should add PHPDoc only when it adds value, not just visual spam. Code should be self explanatory if possible. Like Uncle Bob says: _Software code should read like well-written prose_

File: src/Query/Expression.php (hunk 4 above)
Comment (dorantor) [reply src/Query/Expression.php]: Removed.

File: src/Query/Expression.php (hunk 2 above)
Comment (simPod) [reply src/Query/Expression.php]: > Actually, I can see phpdoc at __constructor for ClickHouseDB\Query\Query. Are you sure it should be removed? :)

Yes, I did not manage to clean it up everywhere yet :(

File: src/Query/Expression.php (hunk 3 above)
Comment (dorantor) [reply src/Query/Expression.php]: Updated docblock with example and param type.

File: src/Query/Expression.php (hunk 1 above)
Comment (dorantor) [reply src/Query/Expression.php]: Done. Sorry, just now understood that this change could be applied right in github.

File: src/Client.php (hunk 5 above)
Comment (simPod) [reply src/Client.php]: Can you specify the added convenience pls?

`$chc->expr("UUIDStringToNum('{$uuid}')");`

vs

`new Expression("UUIDStringToNum('{$uuid}')");`

I'm just trying to find out whether it can be dropped so `Client` class doesn't unnecessarily grow, it already does too much things if you know what I mean ;)

File: src/Query/Expression.php (hunk 1 above)
Comment (simPod) [reply src/Query/Expression.php]: Np, new feature introduced yesterday ;)

File: src/Query/Expression.php (hunk 2 above)
Comment (dorantor) [reply src/Query/Expression.php]: I agree. But I also have a counter for that :) When everything have docblock - it adds rhythm, so code is easer to read, visually separate into blocks(visually extract methods, in this case).
And after I've added example into `__constructor` docblock not sure how to rewrite this block. Just remove `Expression constructor.` line would be good?

File: src/Client.php (hunk 5 above)
Comment (dorantor) [reply src/Client.php]: Just made a comparison for two cases:
```php
    new Expression("UUIDStringToNum('{$uuid}')")
    $chc->expr("UUIDStringToNum('{$uuid}')")
```

Case with `expr()` call is slightly shorter, as you can see(and will be for most people IMHO). But, what is more important it's much more clear what kind of _expression_ is used here. If you're reading this code for the first time it will be more clear what's going on. In my opinion ofc :)

One more for `expr()` - it incapsulates what's really going on inside. 
So, on one side - interface for `Client` is changed, on the other side it hides innerworks and makes it's easer to change them later.

Comment (simPod) [reply src/Client.php]: > Case with expr() call is slightly shorter, as you can see(and will be for most people IMHO)

That's just because abbreviation is used but IMO that's not relevant. You could simply make class `E` and say it's shorter. Abbreviations should not be used if not standardized (like IPAddress -> IP)

> But, what is more important it's much more clear what kind of expression is used here

That's rather determined by namespace -> Query Expression. Client has nothing to do with it I think. Client's responsibility is to make connection to database possible and execute operations.

> it incapsulates what's really going on inside

Encapsulation is done by the `Expression` object already, no need for two ;)

File: src/Query/Expression.php (hunk 2 above)
Comment (simPod) [reply src/Query/Expression.php]: That rythm is there as well when everything has no docblock ;). The visual noise is reduced. It also helps that we have recently integrated code style checks so it all then feels visually unified (it's not applied everywhere yet). Those visually separated blocks are methods themselves already. 

Soon there won't be need for comments at all as typed properties are coming with PHP 7.3 and typed arrays are planned too. So comments will make really sense only if they explain something. Currently they're usually used to specify what type is in arrays.

And yes, removing `Expression constructor` is good enough as the rest of the comment adds value. It helps the developer to understand how to use the class.

Also `@param string $expression` can be removed, that's already in method's signature

**Comment by simPod** [comment]
Travis Code Style build is currently broken (will be fixed when https://github.com/smi2/phpClickHouse/pull/100 is merged) so it can be ignored

Scrutinizer however found some issues, mostly the code style https://scrutinizer-ci.com/g/smi2/phpClickHouse/inspections/e96a39a6-5929-44aa-ae63-943b77da290d/issues/

File: src/Client.php (hunk 5 above)
Comment (simPod) [reply src/Client.php]: BTW some cleanup I've worked on recently https://github.com/smi2/phpClickHouse/pull/102 Trying to clean the API so every class doesn't do everything. The goal is to make public API self-explanatory and not confusing.

Comment (dorantor) [reply src/Client.php]: After some thinking I agree that `expr()` method is not `Client` responsibility. But I still think that `Expression` creation is library responsibility. User can, but should not create `Expression` object directly.

**Comment by dorantor** [comment]
Scrutinizer is now ok. I'll remove method `expr()` from client, but, as I mentioned, still think `Expression` creation should be encapsulated by library. But I'm not familiar enough with the library to make a good suggestion how and where it should be done.

File: src/Client.php (hunk 5 above)
Comment (dorantor) [reply src/Client.php]: I will resolve this conversation, as I agreed with `expr()` removal.

**Comment by simPod** [comment]
Everything that is public = API. And even though you create another method that instantiates it, it will still be perfectly valid to instantiate it on it's own. If you want to encapsulate it, you can create something like expression builder (example https://www.doctrine-project.org/projects/doctrine-collections/en/latest/expression-builder.html).

File: src/Query/Expression.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /** @var string */
+    private $expression;
+
+    public function __construct(string $expression)
+    {
+        $this->expression = $expression;
+    }
+
+    public function __toString(): string
```
Comment (simPod) [inline src/Query/Expression.php]:
```suggestion
    public function __toString() : string
```

File: src/Client.php [hunk 7]
```diff
@@ -6,6 +6,7 @@
 
 use ClickHouseDB\Exception\QueryException;
 use ClickHouseDB\Query\Degeneration;
+use ClickHouseDB\Query\Expression;
```
Comment (simPod) [inline src/Client.php]:
```suggestion
```

File: src/Query/Expression.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<?php
+
```
Comment (simPod) [inline src/Query/Expression.php]:
```suggestion

declare(strict_types=1);
```

**Comment by dorantor** [comment]
Tests are a must. Just wanted to finalize how it will be used first. Regarding ExpressionBuilder - most probably I will not be able to do it in any close future :)

**Comment by dorantor** [comment]
In older version there was no call `$value = ValueFormatter::formatValue($value, false);` in `src/Quote/StrictQuoteLine.php`, so my solution was working as expected. Now it's there and converts `Expression` to string and applies quoting(which I try to avoid with `Expression` class). Currently I see two options how to make it work again:
 - Make `Expression` implementation of `Type` interface and return `$this` in `getValue()`
 - Add check in `ValueFormatter::formatValue` for `Expression`

But both options doesn't look good to me.

**Comment by dorantor** [comment]
Commit that introduced `ValueFormatter` call says 

> Refactor formatting
But it doesn't gives any idea why type conversion was added. W/o knowing reasons for that I cannot offer a good solution.

**Comment by simPod** [comment]
Yup, that cleaned up value formatting code that was spread across several classes into one

You can replicate this part for Expression and should be good https://github.com/smi2/phpClickHouse/commit/5f4fe723c67b01d3074ee71a307e1e67a635977f#diff-8b1652bb2da78b48114bdfa44b1f9b3fR35

File: tests/Query/ExpressionTest.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        $expressionString = "UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')";
+        $preparedValue = FormatLine::Insert([new Expression($expressionString)]);
+
+        $this->assertEquals(
+            $expressionString,
+            $preparedValue
+        );
+    }
+}
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
}

```

File: tests/Query/ExpressionTest.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+            (string) $expr
+        );
+    }
+
+    public function testExpressionValueForInsert()
+    {
+        $expressionString = "UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')";
+        $preparedValue = FormatLine::Insert([new Expression($expressionString)]);
+
+        $this->assertEquals(
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
        self::assertEquals(
```

File: tests/Query/ExpressionTest.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        $expressionString = "UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')";
+        $expr = new Expression($expressionString);
+
+        $this->assertEquals(
+            $expressionString,
+            (string) $expr
+        );
+    }
+
+    public function testExpressionValueForInsert()
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
    public function testExpressionValueForInsert() : void
```

File: tests/Query/ExpressionTest.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+use PHPUnit\Framework\TestCase;
+
+class ExpressionTest extends TestCase
+{
+    public function testToString()
+    {
+        $expressionString = "UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')";
+        $expr = new Expression($expressionString);
+
+        $this->assertEquals(
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
        self::assertEquals(
```

File: tests/Query/ExpressionTest.php [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace ClickHouseDB\Tests\Query;
+
+use ClickHouseDB\Query\Expression;
+use ClickHouseDB\Quote\FormatLine;
+use PHPUnit\Framework\TestCase;
+
+class ExpressionTest extends TestCase
+{
+    public function testToString()
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
    public function testToString() : void
```

File: tests/Query/ExpressionTest.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+declare(strict_types=1);
+
+namespace ClickHouseDB\Tests\Query;
+
+use ClickHouseDB\Query\Expression;
+use ClickHouseDB\Quote\FormatLine;
+use PHPUnit\Framework\TestCase;
+
+class ExpressionTest extends TestCase
```
Comment (simPod) [inline tests/Query/ExpressionTest.php]:
```suggestion
final class ExpressionTest extends TestCase
```

**Comment by dorantor** [comment]
> Yup, that cleaned up value formatting code that was spread across several classes into one
> 
> You can replicate this part for Expression and should be good [5f4fe72#diff-8b1652bb2da78b48114bdfa44b1f9b3fR35](https://github.com/smi2/phpClickHouse/commit/5f4fe723c67b01d3074ee71a307e1e67a635977f#diff-8b1652bb2da78b48114bdfa44b1f9b3fR35)

Not sure I understood correctly. You mean make it this way?
```php
if ($value instanceof Expression) {
```

File: tests/Query/ExpressionTest.php (hunk 10 above)
Comment (dorantor) [reply tests/Query/ExpressionTest.php]: Why `self` is better then `$this`?

File: tests/Query/ExpressionTest.php (hunk 14 above)
Comment (dorantor) [reply tests/Query/ExpressionTest.php]: Is it essential? I mean it doesn't add any value :) And changes nothing.

**Comment by simPod** [comment]
> Not sure I understood correctly. You mean make it this way?

Exactly. And maybe remove `__toString()` and replace with `get()`, `getValue()` or something similar

File: tests/Query/ExpressionTest.php (hunk 14 above)
Comment (simPod) [reply tests/Query/ExpressionTest.php]: I several codebases I met accidentaly extended Test classes instead of TestCase classes (result of huge refactorings, renaming etc.). This simply prevents that so you're safe, your tests won't ever be extended ;)

File: tests/Query/ExpressionTest.php (hunk 10 above)
Comment (simPod) [reply tests/Query/ExpressionTest.php]: I tried to explain it here few days ago https://github.com/webonyx/graphql-php/pull/354#issuecomment-422900157 

TLDR: the method is a static wrapper and therefore should be called like every other static method you call statically, not dynamically.

File: tests/Query/ExpressionTest.php (hunk 14 above)
Comment (dorantor) [reply tests/Query/ExpressionTest.php]: I see. Now it makes sense.

**Comment by dorantor** [comment]
> > Not sure I understood correctly. You mean make it this way?
> 
> Exactly. And maybe remove `__toString()` and replace with `get()`, `getValue()` or something similar

Main reasoning for class wrapper (`Expression`) is to bypass check in `src/Quote/StrictQuoteLine.php`:
```php
if (is_string($value) && $encode) {
```

If replace `__toString()` with `getValue():string` and use it in`ValueFormatter::formatValue()` as:
```php
        if ($value instanceof Expression) {
            return $value->getValue();
        }
```
check I noted before will not be bypassed.

File: tests/Query/ExpressionTest.php (hunk 10 above)
Comment (dorantor) [reply tests/Query/ExpressionTest.php]: Haven't looked in phpunit sources that carefully. I think it came from phpunit docs. Or at least from early ones.
Thanks for link with deep explanation.

**Comment by dorantor** [comment]
This is what I can offer for now, but I don't like this solution, to be honest.

**Comment by simPod** [comment]
What in particular you don't like about it?

File: src/Query/Expression.php [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+namespace ClickHouseDB\Query;
+
+/**
+ * Pass expression "as is" to be sent and executed at server.
+ * P.ex.: `new Expression("UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')");`
+ */
+class Expression
+{
+    /** @var string */
+    private $expression;
```
[the remark this answers is not in the document]
Comment (dorantor) [reply src/Query/Expression.php]: There was different intention and reasoning for `Type` and `Expression`. 
`Expression` was made to bypass quoting. Inspired by other DB-related libs. But, btw, not necessarily in a best way implement this intention. I just found it less obtrusive and easier to merge in future back to my fork.
`Type`, as I can see, introduced for different reason - to allow safely and in a controlled fashion add new, potentially complex types(what about `Date` type, f.ex.?).

While `Expression` seems to be universal here I'm against replacing `Type` with `Expression`, because `Expression` is a last resort. Unsafe one, I must say. Should be used only when there are no other options. In theory, with time, `Expression` could be replaced with some support for functions from library itself. Like `new Function\UUIDStringToNum($uuid)` instead of `new Expression("UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')")`.

**Comment by dorantor** [comment]
> What in particular you don't like about it?

Cannot say. It just smells bad for me. Something is wrong here. 
Validating against class instead of interface? 

But it's working, so I'm not freaking out :)

File: src/Query/Expression.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (simPod) [reply src/Query/Expression.php]: kk, let's just leave it like it is

**Comment by simPod** [comment]
I suppose some interface will emerge soon for that case. Type and Expression could have shared one.

File: src/Query/Expression.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (dorantor) [reply src/Query/Expression.php]: So, maybe, `Expression` could become either a higher level (abstract?) class or even Interface and `Type` can be a descendant to it. Because you're right in one aspect - `Expression` could solve a wider range of requirements including ones that are solved by `Type`. 
But it will require to refactor this part in `src/Quote/StrictQuoteLine.php`:
```php
            $value = ValueFormatter::formatValue($value, false);

            if (is_float($value) || is_int($value)) {
                return (string) $value;
            }

            if (is_string($value) && $encode) {
```
if you want `Expression::getValue()` to return string. Because otherwise string 
`"UUIDStringToNum('0f372656-6a5b-4727-a4c4-f6357775d926')"`
will become 
`'\'UUIDStringToNum(\\\'0f372656-6a5b-4727-a4c4-f6357775d926\\\')\''`
after quoting.

[the remark this answers is not in the document]
Comment (simPod) [reply src/Query/Expression.php]: You're right it might need some tweak in that part of code. But see this line

https://github.com/smi2/phpClickHouse/blob/master/src/Quote/StrictQuoteLine.php#L74

If replaced by `if ($value instanceof Expression) {` it might do the trick.

BTW I'd definitely be for `interface Expression` rather than `abstract class Expression`

[the remark this answers is not in the document]
Comment (dorantor) [reply src/Query/Expression.php]: Yes, missed that one. So it should be something like this IMO:
```php
            if ($value instanceof NumericType) {
                $encode = $value->doQuoting();
            }
```

But in this case abstract class AND interface required. Abstract is required to define default `doQuoting() : boolean`. I think I can do it after this one is merged. Because, obviously, it will be refactoring of `Type` and `Expression`.

[the remark this answers is not in the document]
Comment (simPod) [reply src/Query/Expression.php]: Yes. I'd name it something like `needsEncoding() : bool` and abstract class is IMO not required, we don't really need default value so interface will do. But for sure in next PR :)

**Comment by isublimity** [comment]
Merge to master 
https://github.com/smi2/phpClickHouse/commit/ed7ae42d4b51687cf86ae7a9777f20cb9adc08ec
